unlock-app/src/components/creator/CreatorLockForm.js

```diff
@@ -14,86 +14,157 @@ import {
-      (name === 'maxNumberOfKeys' && value === '∞' && unlimitedKeys) ||
-      (name === 'keyPrice' && !isNaN(value) && value >= 0)
-    )
+    const missing = val => val === undefined || val === null || val === ''
+    const isPositiveInteger = val =>
+      !isNaN(val) && +val === parseInt(val) && +val > 0
+    const isPositiveNumber = val => !isNaN(val) && +val > 0
+    switch (name) {
+      case 'name':
+        if (missing(value)) {
+          return FORM_LOCK_NAME_MISSING
+        }
+        break
+      case 'expirationDuration':
+      case 'maxNumberOfKeys': {
+        const error = {
+          expirationDuration: FORM_EXPIRATION_DURATION_INVALID,
+          maxNumberOfKeys: FORM_MAX_KEYS_INVALID,
+        }[name]
+        if (value !== '∞' && !isPositiveInteger(value)) {
+          return error
+        }
+        break
+      }
+      case 'keyPrice':
+        if (!isPositiveNumber(value)) {
+          return FORM_KEY_PRICE_INVALID
+        }
+        break
+    }
+    return false
   }
 
   handleUnlimitedClick() {
-    this.setState({ unlimitedKeys: true, maxNumberOfKeys: '∞' })
-    // Marking the maxNumberOfKeysRef field as valid.
-    this.maxNumberOfKeysRef.current.dataset.valid = true
+    this.setState(state => ({
+      unlimitedKeys: true,
+      maxNumberOfKeys: '∞',
+      valid: this.valid({ ...state, [name]: '∞' }),
+    }))
   }
 
-  handleChange(event) {
-    let { name, value, dataset } = event.target
-
-    if (name === 'maxNumberOfKeys') {
-      dataset.valid = value === '∞' ? true : this.validate(name, value)
-      this.setState({ unlimitedKeys: value === '∞' })
-    } else {
-      dataset.valid = this.validate(name, value)
-    }
-
-    this.setState({ [name]: value })
+  handleChange({ target: { name, value } }) {
+    this.setState(state => ({
+      unlimitedKeys:
+        name === 'maxNumberOfKeys' ? value === '∞' : state.unlimitedKeys,
+      [name]: value,
+      valid: this.valid({ ...state.valid, [name]: value }),
+    }))
   }
 
   handleSubmit() {
-    if (document.querySelector('[data-valid="false"]')) return false
+    this.setState(state => {
+      const { validityState: valid, errors } = this.valid(state)
+      const { setError, resetError } = this.props
+      const allFormErrors = [
+        FORM_EXPIRATION_DURATION_INVALID,
+        FORM_KEY_PRICE_INVALID,
+        FORM_MAX_KEYS_INVALID,
+        FORM_LOCK_NAME_MISSING,
+      ]
+      allFormErrors.forEach(error => {
+        if (errors.indexOf(error) >= 0) {
+          setError(error)
+        } else {
+          resetError(error)
+        }
+      })
+      if (!valid.formValid) return { valid, errors }
 
-    const { account, createLock, hideAction } = this.props
-    const {
-      expirationDuration,
-      expirationDurationUnit,
-      keyPriceCurrency,
-      maxNumberOfKeys,
-      unlimitedKeys,
-      keyPrice,
-      name,
-    } = this.state
+      const { account, createLock, hideAction } = this.props
+      const {
```

Comment:
Maybe move the part which does create the lock from `state` to its own function to make `handleSubmit` easier to read?